### ISSUE TITLE
Lexer ignoring priority

### ISSUE THREAD

**Issue by TabulateJarl8** [maciejhirsz/logos#297]
Hey! I was thinking of using this library for some lexing work, and I ran into an issue using `logos==0.13.0`. Here's a reduced version of my Token enum:

```rs
#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"[ \t\n\f]+")]
#[logos(subpattern float=r"-?[0-9]+\.[0-9]+")]
pub enum Token {
    // optional negative sign
    // optional float
    // an e followed by an option + or -
    // 1 or more digits (the number following the e)
    // an optional decimal point followed by 1 or more digits (3.1) or (.1)
    #[regex(r"((?&float)?([eE][-+]?\d+(\.\d+)?))", priority=3)]
    ScientificNotation,
    #[regex("(?&float)+", priority=2)]
    Float,
    #[regex(r"-?[0-9]+", priority=1)]
    Int,

    #[regex(r"-", priority=4)]
    Subtract,
}
```

As you can see I'm trying to match negative numbers as well as the subtract sign. The issue is that the lexer is ignoring the priority that I set on the derives and is incorrectly matching things. For example, it should lex `3-3` as `Int, Subtract, Int`, but it is producing `Int, Int`, matching `3` and `-3`. I tried not using subpatterns in case that was the issue, but it didn't seem to fix anything. I feel like I might've seen a related issue pertaining to something similar to this, but it is currently 3AM and I can't seem to find it again. Any help would be greatly appreciated, thanks!

**Comment by maciejhirsz** [comment]
This is working as intended, see [comments starting here](https://github.com/maciejhirsz/logos/issues/173#issuecomment-880798544) in #173.

TL;DR: the lexer will _always_ try to find the longest possible match from its current position, priority is only used if that strategy could possibly match two definitions. Common case is any keyword like `#[token("let")]` vs identifier `#[regex("[a-z]+")]`: `let` should always be a keyword, but `letfoo` should always be an identifier, even though `#[token("let")]` has higher priority than `#[regex("[a-z]+")]`.

In your specific case, if `Substract` was always the valid match for `-`, even if a longer match was available, then the lexer would never produce an `Int` starting with a `-`: it would be a dead pattern.

Typically what you do is only lex for unsigned integers, and only in the AST parser do you construct a signed integer from a unary expression of a `-` (or `+`) operator and and unsigned integer operand. This is true even in Rust itself: `-10` is always two tokens.

**Comment by TabulateJarl8** [comment]
Thank you! I was not aware of this behavior
